zeroforcing/metagraph.pyx

```diff
@@ -0,0 +1,252 @@
+        cdef bitset_t initially_filled_subset
+        cdef bitset_t vxs_to_add
+        
+        bitset_clear(self.filled_set)
+        bitset_clear(self.vertices_to_check)
+        bitset_clear(self.vertices_to_recheck)
+        bitset_clear(self.filled_neighbors)
+        bitset_clear(self.unfilled_neighbors)
+        bitset_clear(self.filled_neighbors_of_vx_to_fill)
+        
+        bitset_union(self.filled_set, &initially_filled_subset2._bitset[0], &vxs_to_add2._bitset[0])
+
+        bitset_copy(self.vertices_to_check, &vxs_to_add2._bitset[0])
+
+        for v in range(self.num_vertices):
+            if bitset_in(&vxs_to_add2._bitset[0], v):
+                bitset_intersection(self.filled_neighbors, self.neighborhood_array[v], self.filled_set)
+                bitset_union(self.vertices_to_check, self.vertices_to_check, self.filled_neighbors)
+            
+        bitset_clear(self.vertices_to_recheck)
+        while not bitset_isempty(self.vertices_to_check):
+            bitset_clear(self.vertices_to_recheck)
+            for vertex in range(self.num_vertices):
+                if bitset_in(self.vertices_to_check, vertex):
+                    bitset_intersection(self.filled_neighbors, self.neighborhood_array[vertex], self.filled_set)
+                    bitset_difference(self.unfilled_neighbors, self.neighborhood_array[vertex], self.filled_neighbors)
+                    
+                    if bitset_len(self.unfilled_neighbors) == 1:
+                        self.vertex_to_fill = bitset_next(self.unfilled_neighbors, 0)
+                        bitset_add(self.vertices_to_recheck, self.vertex_to_fill)
+                        
+                        bitset_intersection(self.filled_neighbors_of_vx_to_fill, self.neighborhood_array[self.vertex_to_fill], self.filled_set)
+                        bitset_remove(self.filled_neighbors_of_vx_to_fill, vertex)
+                        bitset_union(self.vertices_to_recheck, self.vertices_to_recheck, self.filled_neighbors_of_vx_to_fill)
+                        
+                        bitset_add(self.filled_set, self.vertex_to_fill)
+            bitset_copy(self.vertices_to_check, self.vertices_to_recheck)
+
+        self.num_vertices_checked += 1            
+        
+        set_to_return = FrozenBitset(capacity=self.num_vertices)
+        bitset_copy(&set_to_return._bitset[0], self.filled_set)
+        return set_to_return
+    
+
+    cdef neighbors_with_edges_add_to_queue(self, FrozenBitset meta_vertex, FastQueueForBFS queue, int previous_cost):
+        # verify that 'meta_vertex' is actually a subset of the vertices
+        # of self.primal_graph, to be interpreted as the filled subset
+
+        cdef int new_vx_to_make_force
+        cdef int cost
+        cdef int i
+        cdef int num_unfilled_neighbors
+
+        bitset_copy(self.meta_vertex, &meta_vertex._bitset[0])
```

Comment:
Since this is kind of hacky, maybe we should explain it a bit

Reply:
[Cython docs](http://docs.cython.org/en/latest/src/userguide/external_C_code.html#pointers) on this